src/i2c.rs

```diff
@@ -0,0 +1,268 @@
+use crate::callback::CallbackSubscription;
+use crate::callback::Consumer;
+use crate::result::TockResult;
+use crate::syscalls;
+use core::marker::PhantomData;
+use core::ops::Deref;
+use core::ops::DerefMut;
+use libtock_core::shared_memory::SharedMemory;
+
+const DRIVER_NUMBER: usize = 0x20006;
+
+pub const MASTER_BUFFER_SIZE: usize = 32;
+pub const SLAVE_BUFFER_SIZE: usize = 32;
+
+mod command_nr {
+    pub const CHECK_PRESENT: usize = 0;
+    pub const MASTER_WRITE: usize = 1;
+    pub const MASTER_READ: usize = 2;
+    pub const SLAVE_LISTEN: usize = 3;
+    pub const SLAVE_ENABLE_READ: usize = 4;
+    pub const SLAVE_STOP_LISTEN: usize = 5;
+    pub const SLAVE_SET_ADDRESS: usize = 6;
+    pub const MASTER_WRITE_READ: usize = 7;
+}
+
+mod subscribe_nr {
+    pub const SUBSCRIBE_CALLBACK: usize = 0;
+}
+
+mod allow_nr {
+    pub const MASTER_WRIE: usize = 0;
+    pub const MASTER_READ: usize = 1;
+    pub const SLAVE_READ: usize = 2;
+    pub const SLAVE_WRITE: usize = 2;
+}
+
+#[non_exhaustive]
+pub struct I2cDriverFactory;
+
+impl I2cDriverFactory {
+    pub fn init_driver(&mut self) -> TockResult<I2cDriver> {
```

Comment:
Is there a reason init_driver returns a TockResult rather than just I2cDriver? Why not call it `new`?

Reply:
This is just based on what all the other drivers do, I'm just trying to keep everything similar.